snapshot-cassandra/src/main/scala/com/evolutiongaming/kafka/journal/snapshot/cassandra/SnapshotCassandra.scala

```diff
@@ -0,0 +1,172 @@
+    config: SnapshotCassandraConfig,
+    origin: Option[Origin],
+    cassandraClusterOf: CassandraClusterOf[F]
+  ): Resource[F, SnapshotStoreFlat[F]] = {
+
+    def store(implicit cassandraCluster: CassandraCluster[F], cassandraSession: CassandraSession[F]) =
+      of(config.schema, origin, config.consistencyConfig, config.numberOfSnapshots)
+
+    for {
+      cassandraCluster <- CassandraCluster.of[F](config.client, cassandraClusterOf, config.retries)
+      cassandraSession <- cassandraCluster.session
+      store <- store(cassandraCluster, cassandraSession).toResource
+    } yield store
+
+  }
+
+  def of[F[_]: Temporal: Parallel: CassandraCluster: CassandraSession: LogOf](
+    schemaConfig: SnapshotSchemaConfig,
+    origin: Option[Origin],
+    consistencyConfig: CassandraConsistencyConfig,
+    numberOfSnapshots: Int
+  ): F[SnapshotStoreFlat[F]] =
+    for {
+      schema <- SetupSnapshotSchema[F](schemaConfig, origin, consistencyConfig)
+      statements <- Statements.of[F](schema, consistencyConfig)
+    } yield SnapshotCassandra(statements, numberOfSnapshots)
+
+  private sealed abstract class Main
+
+  def apply[F[_]: MonadThrow](statements: Statements[F], numberOfSnapshots: Int): SnapshotStoreFlat[F] = {
+    new Main with SnapshotStoreFlat[F] {
+
+      def save(key: Key, snapshot: SnapshotRecord[EventualPayloadAndType]): F[Unit] = {
+        statements.selectMetadata(key).flatMap {
+          // such snapshot is already saved, overwrite
+          case s if s.values.exists { case (seqNr, _) => snapshot.snapshot.seqNr == seqNr } =>
+            update(key, s, snapshot)
+          // there is a free place to add a snapshot
+          case s if s.size < numberOfSnapshots => insert(key, s, snapshot)
+          // all rows are taken, we have to update one of them
+          case s => replace(key, s, snapshot)
+        }
+      }
+
+      def insert(
+        key: Key,
+        savedSnapshots: Map[BufferNr, (SeqNr, Instant)],
+        snapshot: SnapshotRecord[EventualPayloadAndType]
+      ): F[Unit] = {
+        val allBufferNrs = BufferNr.listOf(numberOfSnapshots)
+        val takenBufferNrs = savedSnapshots.keySet
+        val freeBufferNr = allBufferNrs.find(bufferNr => !takenBufferNrs.contains(bufferNr))
+        MonadThrow[F].fromOption(freeBufferNr, SnapshotStoreError("Could not find a free key")).flatMap { bufferNr =>
+          val wasApplied = statements.insertRecord(key, bufferNr, snapshot)
+          wasApplied.flatMap { wasApplied =>
+            // TODO: consider adding circuit breaker here
+            if (wasApplied) ().pure[F] else save(key, snapshot)
+          }
+        }
+      }
+
+      def replace(
+        key: Key,
+        savedSnapshots: Map[BufferNr, (SeqNr, Instant)],
+        insertSnapshot: SnapshotRecord[EventualPayloadAndType]
+      ): F[Unit] = {
+        val oldestSnapshot = savedSnapshots.toList.minByOption { case (_, (seqNr, timestamp)) => (seqNr, timestamp) }
```

Comment:
Are you expecting to have few snapshots with same `seqNr`? I do not understand use-case for it, could you please share more details

Reply:
It should never have several snapshots with the same `seqNr`, but it seemed like a good idea to be a little bit defensive here.

I.e., if there is a bug or something, it will still filter by `timestamp` and delete an oldest one, not a random one.

Reply:
Would it be possible to log (or report any other way) the case of duplicated `seqNr`? Please resolve the conversation if its not easily possible